---
title: "Detection of the signal of selection from genome scan. Population based analysis "
---

# Introduction

The purpose of the vignette is to detect outlier SNP loci from a large number of
SNPs genotyped for a large number of individuals. Individuals are organised in
well defined populations. Then, we tested if allele counts in outlier loci are
correlated to the temperature. We will first use a method based on principal
component analysis (*PCAdapt*) (section 2) and a second method based on atypical
values of Fst (*OutFLANK*; Whitlok and Lotterhos (2015);
https://github.com/whitlock/OutFLANK) (section 3) to detect outliers loci. Then,
we will test if the allele counts in outliers are correlated to the temperature
using a Poisson regression (section 4).

 

# Assumptions

- We assume that loci are neutral and we are looking for those are deviant from
the neutral assumption.
- We assume that loci correlated to the environment (temperature) are under
selection.

# Data

We will use simulated data provided by Katie Lotterhos (Lotterhos et al. in
prep). The initial data set is constituted of 509 individuals, 19 populations
and 3084 SNP. It has 2 extra-columns: Population and Temperature measures. SNP
are in columns: each row contains one value at each SNP locus (separated by
spaces or tabulations) corresponding to the number of reference alleles: 0 means
zero copies of the reference allele, 1 means one copy of the reference allele
(i.e. heterozygote), and 2 means two copies of the reference allele.


# Resources/Packages required

- *PCAdapt*: http://membres-timc.imag.fr/Michael.Blum/PCAdapt.html
- *OutFLANK*: https://github.com/whitlock/OutFLANK

Since *OutFLANK* is not yet on CRAN, you will need to install it from GitHub
using the *devtools* R package. Additionally, *OutFLANK* depends on *qvalue*
from
[Bioconductor](http://www.bioconductor.org/packages/release/bioc/html/qvalue.html),
which must be installed beforehand.

```{r, packages, eval = FALSE}
install.packages("devtools")
library("devtools")
source("https://bioconductor.org/biocLite.R")
biocLite("qvalue")
install_github("whitlock/OutFLANK")
```

Loading the required packages:

```{r, loadpackage, message = FALSE}
library("pcadapt")
library("qvalue")
library("OutFLANK")
library("ggplot2")
```

 
## Section 1: Load the data

```{r, load_data_show, eval=FALSE}
sel <- read.table("SNPselection1.txt", head = TRUE)
dim(sel)
```

```{r, load_data_evaluate, echo=FALSE}
sel <- read.table("../data/SNPselection1.txt", head = TRUE)
dim(sel)
```

## Section 2: *PCAdapt*

To run the function `pcadapt()`, the user should specify the number K of
principal components (PC) to work with: first perform with a large number of
principal components (e.g. higher than the number of populations), then use the
'scree plot' to chose the value of K. It displays the percentage of variance
that is explained by each PC. The recommended value of K corresponds to the
largest value of K before the plateau of 'scree plot' is reached. Then for a
given SNP, a statistical test to define the SNP as outlier or not is based on
the "loadings" that are defined as the correlation between the SNP and the
PCs. The statistic test for detecting outlier SNPs is the Mahalanobis distance
(between the K correlations of the SNP and each axis and mean correlations) and,
which scaled by a constant, should have a chi-square distribution with K degrees
of freedom under the assumption that there are no outlier. By default P-values
of SNPs with a minor allele frequency smaller than 0.05 are not computed. A
Manhattan plot displays log10 of the p-values. It is also possible to check the
distribution of the p-values using a Q-Q plot. The authors suggest to use false
discovery rate (q-value) to provide a list of outliers.

> Note: *PCAdapt* expects the incoming matrix to have samples in columns and
> loci in rows. Because our data is the opposite, we need to tell the `pcadapt()`
> function to transpose the matrix before analysis with `transpose = TRUE`.

```{r, PCAdapt}

genotype <- sel[, 3:ncol(sel)]
dim(genotype)
K <- 25
x <- pcadapt(genotype, K = K, transpose = TRUE)
plot(x, option = "screeplot") # 19 groups seems to be the correct value
plot(x, option = "scores", pop = sel[, 1]) # how populations are shared among the 19 groups

K <- 19
x <- pcadapt(genotype, K = K, min.maf = 0, transpose = TRUE)

summary(x) # numerical quantities obtained after performing a PCA
plot(x, option = "manhattan")
plot(x, option = "qqplot", threshold = 0.1)
plot(x, option = "stat.distribution") # Distribution of Mahalanobis distances.

qval <- qvalue(x$pvalues)$qvalues
alpha <- 0.1
outliers_pcadapt <- which(qval < alpha)
print(outliers_pcadapt)
length(outliers_pcadapt) # 14 outliers

alpha <- 0.05 # use of a more stringent threshold to detect outliers
outliers <- which(qval < alpha)
print(outliers)
length(outliers) # 14 outliers

```


##section 3: *OutFLANK* (https://github.com/whitlock/OutFLANK)

A procedure to find Fst outliers based on an inferred distribution of neutral
Fst: it uses likelihood on a trimmed distribution of Fst values to infer the
distribution of Fst for neutral markers. This distribution is used to assign
q-values to each locus to detect outliers loci potentially due to spatially
heterogeneous selection. In practice, first the function `MakeDiploidFSTMat()`
allows to calculate the appropriate input data frame among other parameters, Fst
for each locus, and Fst without sampling size correction (`FSTNoCorr`). Then the
function `OutFLANK()` estimate q-values and provides a list of outliers.

```{r, outFLank}

pop <- data.frame(sel[, 1])
ind <- paste("pop", pop[, 1]) #vector with the name of population

locinames <-
as.character(seq(1:ncol(genotype))) #vector with the name of loci

FstDataFrame <- MakeDiploidFSTMat(genotype, locinames, ind)
plot(FstDataFrame$FST, FstDataFrame$FSTNoCorr, xlim = c(-0.01,0.3), 
     ylim = c(-0.01, 0.3), pch = 20)
abline(0, 1) # Checking the effect of sample size on Fst since FSTCoCorr will be used in the follow
hist(FstDataFrame$FSTNoCorr) 

OF <- OutFLANK(FstDataFrame, NumberOfSamples=19, qthreshold = 0.05, 
               RightTrimFraction = 0.05)

# Plot the ditribution of Fst with the chi squared distribution
OutFLANKResultsPlotter(OF, withOutliers = TRUE, NoCorr = TRUE, Hmin = 0.1, 
                       binwidth = 0.005, Zoom = FALSE, RightZoomFraction = 0.05, 
                       titletext = NULL)


outliers_OF <- OF$results$LocusName[OF$results$OutlierFlag == TRUE]
print(outliers_OF)
length(outliers_OF) # 11 outliers

```


## Section 4: Logistic regression : linking outliers and temperature

We aim to test if the outliers detected with *PCAdapt* and *OutFLANK* (663 706
923 1163 1378 1639 1666 1825 2133 2556 2871) are correlated to the temperature.
We use a GLM with a binomial error to model the probability to have the allele
"A" out of the two alleles. We derive the relation for some of the shared
outliers. You can do the test for all.


```{r,Binomial regressions}

# We keep the outlier loci selected by both pcadapt and outflank
outliers <- outliers_pcadapt[outliers_pcadapt %in% outliers_OF]
length(outliers) # 11 outliers

loc1 <- genotype[, outliers[1]]
temp <- sel$Temperature
loc1temp <- data.frame(loc1, temp)
 
mod <- glm(cbind(loc1, 2 - loc1) ~ temp, family = binomial) 
summary(mod) # This locus is significantly correlated to temperature

 
loc2 <- genotype[, outliers[2]]
loc2temp <- data.frame(loc2, temp)
ggplot(loc2temp, aes(x = factor(loc2), y = temp)) + 
 geom_boxplot() + 
 xlab("Major allele count") +
 ylab("Temperature (Centigrade)")


mod <- glm(cbind(loc2, 2 - loc2) ~ temp, family = binomial) 
summary(mod) # This locus is significantly correlated to temperature

# Test with a locus that is only selected by one method
loc7 <- genotype[, outliers_pcadapt[!outliers_pcadapt %in% outliers_OF][1]]
loc7temp <- data.frame(loc7, temp)
  
ggplot(loc2temp, aes(x = factor(loc7), y = temp)) + 
 geom_boxplot() + 
 xlab("Major allele count") +
 ylab("Temperature (Centigrade)")

mod <- glm(cbind(loc7, 2 - loc7) ~ temp, family = binomial)
summary(mod) ## This locus is not significantly correlated to temperature
 
```

# Conclusions

- *PCAdapt* detected 14 outliers 
- *OutFLANK* detected 11 outliers 
- 11 outliers were detected in common 
- Those common outliers are all significantly correlated with the temperatures
as shown by Poisson regression. Such outliers can be potentially involved in
local adaptation due to temperatures. However more analysis are necessary to
confirm that they are outliers, and which allele is potentially under selection.



## What's next

Information on further analysis that could be done as LFFM (package *LEA*..).

# Contributors

- Stephanie Manel (EPHE) (Author) 
- Alicia Dalongeville (CEFE) (Author)
- Zhian Kamvar (reviewer)

#References

Duforet-Frebourg N, K Luu, G Laval, E Bazin, MGB Blum. (2016) Detecting genomic
signatures of natural selection with principal component analysis: application
to the 1000 Genomes data. Molecular Biology and Evolution
(http://arxiv.org/abs/1504.04543)

Whitlock MC, Lotterhos KE (2015) Reliable Detection of Loci Responsible for
Local Adaptation: Inference of a Null Model through Trimming the Distribution of
Fst. The American Naturalist 186, S24-S36.
http://www.jstor.org/stable/10.1086/682949

# Session Information

This shows us useful information for reproducibility. Of particular importance
are the versions of R and the packages used to create this workflow. It is
considered good practice to record this information with every analysis.

```{r, sessioninfo}

options(width = 100)
devtools::session_info()

```



