---
title: "Population Differentiation for Sequence Data"
---

# Introduction

Now, we will discuss in more detail how to assess population genetic structure from sequence data. Assuming that you have a priori information about how the individuals are grouped in subpopulations, one can do the following analyses: 1) quantify pairwise subpopulation structure and their significance, 2) test for hierarchical structure among groups of subpopulations, and 3) use population clustering algorithms to corroborate the a priori grouping hypothesis. We will go into each of these analyses in this vignette.  

# Resources

## Data

Data used for the following analyses come from one mitochondrial gene (Cytochrome Oxidase I = ['Ebom_mt.fas'](https://github.com/nescent/popgenInfo/blob/master/data/Ebom_mt.fas)) and one nuclear gene (Calcium/Calmodulin-Dependent Protein Kinase II = ['Ebom_CAD.fas'](https://github.com/nescent/popgenInfo/blob/master/data/Ebom_CAD.fas)) of the orchid bee *Eulaema bombiformis*.

## Packages

```{r, packages, message = FALSE}
library("apex")
library("adegenet")
library("pegas")
library("mmod")
library("poppr")
```

# Importing data

We will first import FASTA files for all genes at once using the *apex* function `read.multiFASTA()`. This will allow us to read in multiple FASTA formatted files at once into a "multiDNA" object from the *apex* package. We can plot it to see what the two genes look like for all of our samples

```{r data_import_FASTA, eval = FALSE}
# Creating DNAbin objects
beeData <- read.multiFASTA(c("Ebom_mt.fas", "Ebom_CAD.fas"))
plot(beeData, cex = 0.2)
```

```{r data_import_FASTA_real, echo = FALSE}
beeData <- read.multiFASTA(c("../data/Ebom_mt.fas", "../data/Ebom_CAD.fas"))
plot(beeData, cex = 0.2)
```

> Pro tip: You can list all the files ending in ".fas" in your folder by using 
> `dir(pattern = "fas", full.names = TRUE)`

Now, this "multidna" object (`beeData`) will be converted into a "genind" object
(`beeData.gid`), which will be used for downstream analyses. One thing we need
to be aware of is the fact that we can't have any periods in our locus names. By
default, the locus names are set to the file names, which both contain ".fas" in
the title. We can fix this by using the text-replacing function `gsub()`:

```{r fix_loci}
getLocusNames(beeData)
(setLocusNames(beeData) <- gsub(".fas", "", getLocusNames(beeData)))
```

Now we can create our "genind" object.

```{r, data_manipulate_FASTA}

# Creating genind object by multilocus sequence types
beeData.gid <- multidna2genind(beeData, mlst = TRUE)
beeData.gid
```

We also want to set the population strata.

```{r set_populations}
my_strata <- data.frame(regions = rep(c("West", "East"), each = 20), 
                        populations = rep(c("CA", "Ch", "Am", "AF"), each = 10))
strata(beeData.gid) <- my_strata
setPop(beeData.gid) <- ~populations
beeData.gid
```


# Pairwise population differentiation

## Overall F-statistics

```{r, Fstats1}
diff_stats(beeData.gid) # this function calculates overall Nei's Gst, Hedrick's Gst and  of the dataset
Phi_st_Meirmans(beeData.gid) # this function calculates overall PhiST, the Fst analog for DNA sequence data
```
## Pairwise Fst

```{r, Fstats2}
pairwise_Gst_Nei(beeData.gid, linearized = FALSE) # Calculates pairwise Gst. If linearized = TRUE, it calculates 1/(1- Gst)  
pairwise_Gst_Hedrick(beeData.gid, linearized = FALSE)# Calculates pairwise Gst. If linearized = TRUE, it calculates 1/(1- Gst')  
pairwise_D(beeData.gid, linearized = FALSE, hsht_mean = "harmonic") # Calculates pairwise Gst. If linearized = TRUE, it calculates 1/(1- D)  
```
## Testing for significance

To estimate if populations are significantly different, we will generate 100 replicates of the dataset using the function `chao_bootstrap()`. Then, summary statistics (mean and 95% CI) will be calculated for each of the different parameters of population differentiation. 

```{r, Fstats3}
bs <- chao_bootstrap(beeData.gid, nreps = 100)
summarise_bootstrap(bs, Gst_Nei)     # for Nei's Gst
summarise_bootstrap(bs, Gst_Hedrick) # for Hedrick's Gst
summarise_bootstrap(bs, D_Jost)      # for Jost's D
```

# AMOVA (Analysis of Molecular Variance)

Analysis of Molecular Variance (AMOVA) is a method for estimating population differentiation from molecular data taking into account the mutational distance between alleles. Unlike $F_{st}$, which quantifies genetic differentiation based on allele frequencies, AMOVA treats molecular data as vectors and estimates Euclidean distances between alleles. Furthermore, it is possible to test hypotheses about differentiation by grouping subpopulations in a hierarchical structure (Excoffier et al., 1992).

```{r, eval = TRUE}
beeData_dist <- dist.multidna(beeData, pool = TRUE)
amova(beeData_dist ~ populations, data = strata(beeData.gid), nperm = 100)
```

# Conclusions

In this vignette, we learned how to estimate different parameters of overall and pairwise population differentiation (Nei's Gst, Hedrick's Gst, Jost D) from DNA sequence data. We also learned how to do a hierarchical analysis of population structure using AMOVA.

## What's next

With these exploratory analyses, we understand how genetic diversity is distributed among populations. One could now move to testing specific hypotheses than could explain the patterns of population structure found with the F-statistics.

# Contributors

- Margarita M. López- Uribe (Author)
- Zhian N. Kamvar (edits)

# References

Excoffier L., Smouse PE., Quattro JM. 1992. Analysis of molecular variance inferred from metric distances among dNA haplotypes: Application to human mitochondrial dNA restriction data. Genetics 131:479–491. Available at: http://www.genetics.org/content/131/2/479.abstract

# Session Information

This shows us useful information for reproducibility. Of particular importance are the versions of R and the packages used to create this workflow. It is considered good practice to record this information with every analysis.

```{r, sessioninfo}
options(width = 100)
devtools::session_info()
```
