---
title: "Calculating Basic Population Genetic Statistics from Microsatellite Data"
---

# Introduction

In this vignette, you will calculate basic population genetic statistics from microsatellite data using R packages. 

1. Genetic diversity,
2. Test Hardy Weinberg 
3. $F_{is}$ and global $F_{st}$

# Assumptions

- $H_0$: we assume that the microsatellites are in Hardy-Weinberg equilibrium
- $H_1$: they are not.



# Data

The data used for these analyses are contained in an R dataset: `nancycats`, a "genind" object, i.e. an object of the package *adegenet*.
It contains microsatellite genotypes of 237 cats from 17 colonies of Nancy (France). Each individuals of the 17 colonies are genotyped for 9 microsatellite loci.  


# Resources/Packages required

```{r, packages, message = FALSE}
library("adegenet")
library("pegas")
library("hierfstat")
```

# Analysis

## Section 1: Load the data

```{r, load_data_show}
data("nancycats", package = "adegenet")  
summary(nancycats)
```

## Section 2: Genetic diversity (observed and expected heterozygosity)   


```{r, Genetic_diversity}
nancycats  
div <- summary(nancycats)
div
plot(div$Hobs, xlab="Loci number", ylab="Observed Heterozygosity", 
	 main="Observed heterozygosity per locus")

plot(div$Hobs, div$Hexp, xlab="Observed Heterozygosity", ylab="Expected Heterozygosity", 
     main="Expected heterozygosity as a function of observed heterozygosity per locus")

bartlett.test(list(div$Hexp, div$Hobs)) # difference of means 
```

We get various information from the command summary. 
The one that interest us is the observed and expected heterozygosity per locus.
We observed that heterozygosity varies among loci.

The results from the Bartlett test indicates that we have no difference between the mean observed and expected heterozygosity

## Section 3: Basic statistics with *hierfstat* 

The function `basic.stats()` provides the observed heterozygosity ($H_o$), mean
gene diversities within population ($H_s$), $F_{is}$, and $F_{st}$. The function
`boot.ppfis()` provides confidence interval for $F_{is}$. The function `indpca()`
does a PCA on the centered matrix of individuals' allele frequencies.

```{r, basic_statistics}
nancycats.hfstat <- genind2hierfstat(nancycats)
basicstat <- basic.stats(nancycats, diploid = TRUE, digits = 2) 
names(basicstat)
boot.ppfis(nancycats.hfstat) 
x <- indpca(nancycats.hfstat) 
plot(x, cex = 0.7)
```

## Section 4: Testing for Hardy-Weinberg Equilibrium

```{r, Hardy-Weinberg_test}
hw.test(nancycats, B = 1000)
```

We get for each locus a test of significance of the null hypothesis: $H_0$ the
locus is in HW equilibrium in the population/ $H_1$. The locus is not in HW equilibrium.

We can conclude from the p-values of each test that any locus is in HW equilibrium.

# Conclusions

## What did we learn today?

In this vignette, we learned how to explore the patterns of genetic diversity in one population.
Also, you have an idea of potential violations of the dataset to the null Wright-Fischer model.

## What is next?

You may now want to move on to looking into [population differentiation](2015-12-15-microsatellite-differentiation.html).


# Contributors

- StÃ©phanie Manel (Author)
- Zhian Kamvar (edits)
 

# Session Information

This shows us useful information for reproducibility. Of particular importance are the versions of R and the packages used to create this workflow. It is considered good practice to record this information with every analysis.

```{r, sessioninfo}
options(width = 100)
devtools::session_info()
```



